keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1300 @@
+topology needs to look like at sandbox creation time.
+
+Updating resources of a container means also resizing the VM, hence the
+hypervisors needs the complete list of resources available at a update container
+request.
+
+Another visibility issue is related to the native and extended resources.
+Kubelet manages the native resources (CPU and memory) and communicates resource
+parameters over the CRI API to the runtime. The following snippet shows the
+currently supported CRI annotations that are provided by the Kubelet to e.g.
+`containerd`:
+
+```sh
+pkg/cri/annotations/annotations.go
+
+  // SandboxCPU annotations are based on the initial CPU configuration for the sandbox. This is calculated as the
+  // sum of container CPU resources, optionally provided by Kubelet (introduced  in 1.23) as part of the PodSandboxConfig
+  SandboxCPUPeriod = "io.kubernetes.cri.sandbox-cpu-period"
+  SandboxCPUQuota  = "io.kubernetes.cri.sandbox-cpu-quota"
+  SandboxCPUShares = "io.kubernetes.cri.sandbox-cpu-shares"
+
+  // SandboxMemory is the initial amount of memory associated with this sandbox. This is calculated as the sum
+  // of container memory, optionally provided by Kubelet (introduced in 1.23) as part of the PodSandboxConfig.
+  SandboxMem = "io.kubernetes.cri.sandbox-memory"
+```
+
+However, the original details of
+the resource spec are lost as they get translated (within kubelet) to
+platform-specific (i.e. Linux or Windows) resource controller parameters like
+cpu shares, memory limits etc. Non-native resources such as extended resources
+and the device plugin resources completely invisible to the CRI runtime. However,
+[OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md),
+[runC](https://github.com/opencontainers/runc) wrappers,
+[NRI](https://github.com/containerd/nri) plugins or in some cases even
+applications themselves would benefit on seeing the original resource requests
+and limits e.g. for doing customized resource optimization.
+
+Extending the CRI API to communicate all resources already at sandbox creation
+and pass down resource requests and limits (of native and extended resources)
+would provide a comprehensive and early-enough view of the resource usage of
+all containers of the pod, allowing improved resource allocation without
+breaking any existing use cases.
+
+### Goals
+
+- make the information about all required resources (e.g. native and extended
+  resources, devices, mounts, CDI devices) of a Pod available to the CRI at
+  sandbox creation time
+- make container resource spec transparently visible to CRI (the container
+  runtime)
+
+### Non-Goals
+
+- change kubelet resource management
+- change existing behavior of CRI
+- add UpdatePodSandboxResources CRI rpc (this is covered by [KEP-1287][kep-1287], [PR][kep-1287-beta-pr])
+- add pod-level resource requirements (this is covered by [KEP-2837][kep-2837], [PR][kep-2837-alpha-pr])
```

Comment:
Can you please clarify what does those two statements mean? Do you mean that those KEPs will need to contribute to this API?

The `1287` is targeting beta, while this one is targeting alpha. So this KEP will need to clarify how it will work with InPlace updates

cc: @tallclair @ndixita

Reply:
I could also remove these from non-goals. I added these here because we talk about UpdatePodSandboxResources message and the pod level resource requirements later on in the proposal - I wanted to clarify that these additions are not part of this proposal (KEP-4112).

`UpdatePodSandboxResources` is not yet in the CRI API. KEP-1287 proposes to add that in Beta. So we'd want to sync with that

WDYT @SergeyKanzhelev ?

Reply:
I mostly want to make sure we are not making changes that are not compatible with these two KEPs. So listing them as non-goals sounds like a bad idea. Maybe not a goal for alpha, but definitely a goal for beta.

Reply:
That's certainly what I want, too. To be in-sync between all the KEPs. I tried to state that this KEP (4112)
- is NOT proposing to add the new `UpdatePodSandboxResources` CRI rpc
- is NOT proposing to add pod-level resource requirements
- is proposing to align with those features when they land

The design details section contains details what changes (to align) would be made.

I now dropped these two non-goals as they cause more confusion than clarity. Maybe, instead, we should add in Goals to align with them?